### DISCUSSION TITLE
mongodb: self referencing embedded attributes can't be searched

### DISCUSSION THREAD

**Discussion by lod911** [mikro-orm/mikro-orm#5544] [General]
I'm not sure if I'm doing it correctly. I want to save a infinite hierarchy. To do this, I define an entity with an attribute that refers to itself and as an array.

It is saved correctly (I followed this discussion: #3145 ). But something is wrong with the query.

Attached is the complete test. [embedded.test.ts.txt](https://github.com/mikro-orm/mikro-orm/files/15223520/embedded.test.ts.txt)

```
   // searches
    const f4 = await orm.em.find(EntityEmbeddedParent, {
      rootOfEmbedded: { myEmbedded: { myEmbedded: { name: 'level010101' } } },
    },
  );

    const f5 = await orm.em.find(EntityEmbeddedParent, {
      rootOfEmbedded: { myEmbedded: { myEmbedded: { myEmbedded: { name: 'level01010201' } } } },
    },
    // does also not help -> Execution-Failure
    { populate: ['rootOfEmbedded.myEmbedded.myEmbedded.myEmbedded'] },
  );
```
The problem:
```
    // log f4
    // db.getCollection('entity-embedded-parent').find(
    // { 'rootOfEmbedded.myEmbedded': { myEmbedded: { name: 'level010101' } } }, {}).toArray();

    // -> Should be
    // { "rootOfEmbedded.myEmbedded.myEmbedded.name": "level010101"}

    // log f5
    // db.getCollection('entity-embedded-parent').find(
    // { 'rootOfEmbedded.myEmbedded': { myEmbedded: { myEmbedded: { name: 'level01010201' } } } }, {}).toArray();

    // -> Should be
    // { "rootOfEmbedded.myEmbedded.myEmbedded.myEmbedded.name": "level01010201"}
```

Mongo-Entity on the Database:
![Bildschirmfoto vom 2024-05-06 18-32-52](https://github.com/mikro-orm/mikro-orm/assets/38237424/0fdc6667-e05d-4bd7-add2-ff3a76e59475)

At the moment I solved the problem with jsonpath:
```
      const retJp = jsonpath.query(
        a,
        '$..[?(@.name=="level020102")]',
      ) as EntityEmbeddedParent[];
```
But this is not sexy and not typed.

I saw in #3145 the support for define it will probably never going be supported. But it works! Except querying it.

I also know, there would be other solutions to save infinite hierarchy e.g. saving a parent-key on each entity. But the advantage of storing it as object in mongodb is the quick and easy access of parts of the hierarchy with a simple query.

Does this statement still apply? And also for find? :pleading_face: 

Regards

What driver are you using?
@mikro-orm/mongodb

MikroORM version
6.2.5

Node.js version
v20.12.2

Operating system
Ubuntu 22.04.4 LTS

**Comment by B4nan** [comment]
Your problem is that you are not defining the array as a JSON property, this makes your test pass:

```ts
@Property({ type: 'json' })
myEmbedded?: EntityEmbeddable[] = [];
```

Note that this way you will get POJOs in there, they wont be instances of the `EntityEmbedded` class.

**Comment by lod911** [comment]
Love it! Works like a charm. Thanks!
